pandas/core/reshape/merge.py

```diff
@@ -906,13 +906,20 @@ def _maybe_coerce_merge_keys(self):
                 continue
 
             # if we are numeric, then allow differing
-            # kinds to proceed, eg. int64 and int8
+            # kinds to proceed, eg. int64 and int8, int and float
             # further if we are object, but we infer to
             # the same, then proceed
             if is_numeric_dtype(lk) and is_numeric_dtype(rk):
                 if lk.dtype.kind == rk.dtype.kind:
                     continue
 
+                # check whether ints and floats
+                if is_integer_dtype(rk) and is_float_dtype(lk):
```

Comment:
I am a little leary of merging integer & float columns when the floats are not actually == ints. A warning would be good here if 

``not (lk == lk.astype(rk)).all()`` is True (and the reverse for 2nd case)